### ISSUE TITLE
Grouping: groupby-like operations

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#575]
Have a look at http://xarray.pydata.org/en/stable/groupby.html. We need something similar in scipp.

A main application for neutron data would be grouping of detectors into spectra, or grouping in an algorithm like Mantid's `DiffractionFocussing`. The core of xarray's `groupby` can do mostly what we need. Examples:

**1. Grouping pixels into spectra**

```python
d = d.groupby('spectrum_number').mean(Dim.Detector)
```
What works:
- Would take mean of positions (detector pos -> spectrum pos).

Caveats:
- Output dimension in `xarray` is `spectrum_number` (with a new coordinate). We would prefer to have the position as a dimension and primary coordinate, spectrum numbers should just be labels.
- Any coords that depend on `Dim.Detector` are removed, that is positions etc. would need to be data.
- The does not seem to be a way to retain the original detector IDs in form of a list/set at each output spectrum number.

**2. Grouping pixels as part of diffraction-focussing**

```python
focussed = d.groupby('group').sum(Dim.Position)
```
What works:
- Natural way to apply grouping, output will have same number of spectra as groups.

Caveats:
- Mantids group information is actually based on detector IDs and there is some specialized code for checking that the spectrum grouping matches the diffraction grouping. So actually we would need something like
  ```python
  # assuming we have a column containing the spectrum number for each detector, can it be stored in a dataset?
  # assuming we have a function `same` that checks that the value is constant and then returns it
  d.labels['group'] = grouping.groupby('spectrum_number').same(Dim.Detector)
  focussed = d.groupby('group').sum(Dim.Position)
  ```

**Comment by SimonHeybrock** [comment]
> `# assuming we have a column containing the spectrum number for each detector, can it be stored in a dataset?`

I think this should/could be part of the detector info. There are still gaps (see below), but putting the pieces together we have now:

```python
detector_info = sc.Dataset()
detector_info.coords[Dim.Position] = sc.Variable([Dim.Position], values=...)
detector_info.labels['spectrum_number'] = sc.Variable([Dim.Position], values=...) # multiple dets can have same spec

d = detector_info.groupby('spectrum_number').mean(Dim.Position) # see 1. below
d.labels['detector_info'] = detector_info

group = sc.neutron.diffraction.load('grouping.cal')
d.labels['group'] = grouping.groupby('spectrum_number').same(Dim.Detector)
focussed = d.groupby('group').sum(Dim.Position)
```

Gaps:
1. `groupby` and `mean` (or similar) would actually drop the position coordinate and not operate on it. Is there a clean way to make it into a coordinate?
2. In general (as you can see above), it is not clear where to use `Dim.Position`, `Dim.Detector`, and `Dim.Spectrum`, and how to translate between them in a natural and implicit way.

**Comment by SimonHeybrock** [comment]
More thoughts:
1. Never use `Dim.Detector` or `Dim.Spectrum`. Reason: This should be entirely redundant information (and can actually be insufficient if the same detector can move to different positions). Always use `Dim.Position`.
   - Is this a good reasoning?
   - Can we argue that `Dim.Position` is also bad, because we may have different rotations of the same detector? Why make position special?
2. Regarding the problem that functions like `mean` do not operate on coords or labels: Can we avoid this by just using a special function that *does*? But do we end up in special case handling again?

**Comment by SimonHeybrock** [comment]
Technical challenge: Due to type-erasure it is not so easy to create/define the "groupby" object, which is essentially a map from the grouped label to a dataset. Either we need to template it (making it inconvenient to use?), or maybe use a `DataArray` as an output, with data elements that are the group datasets? The latter would lack the map-like interface but this might just be a minor disadvantage.

**Comment by SimonHeybrock** [comment]
# Proposed solution for the API-side
(technical details not clear yet!)

I see two main cases, which we can consider by example.

# 1. Theta binning

Typically we start of with data for every detector pixel, but want to reduce this to some derived parameter, e.g., the scattering angle (theta).

```python
# start with standard dataset
d = sc.Dataset()
d.coords[Dim.Position] = sc.Variable([Dim.Position], values=...)
d.labels['component_info'] = component_info
d['sample'] = sc.Variable([Dim.Position, Dim.Tof], values=...)
d['vanadium'] = sc.Variable([Dim.Position, Dim.Tof], values=...)

# compute two_theta information
d.labels['two_theta'] = sc.neutron.two_theta(d)

# bin into two_theta
two_theta_bins = sc.Variable([Dim.TwoTheta], values=[0,pi/2,pi])
d = d.groupby('two_theta', bins=two_theta_bins).sum(Dim.Position)
```

The resulting `d` looks as follows:
- `Dim.Position` is gone (due to `sum`).
- `Dim.TwoTheta` is a new dimension, with bin-edge coord `two_theta_bins`. This is added in the (implicit) "combine" step.

[WIP]

**Comment by SimonHeybrock** [comment]
Basics have been implemented in #624 and #635. We need more cleanup of the API, but the discussion can continue elsewhere.
